File: python/Q230_kth_smallest_element_in_a_bst.py
#
# @lc app=leetcode id=230 lang=python3
#
# [230] Kth Smallest Element in a BST
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def inorder_with_max(self, root: TreeNode, k: int) -> None:
        if len(self.values) >= k:
            return

        if root.left:
            self.inorder_with_max(root.left, k)

        self.values.append(root.val)

        if root.right:
            self.inorder_with_max(root.right, k)

    def kthSmallest(self, root: TreeNode | None, k: int) -> int:
        self.values: list[int] = []

        assert root is not None
        self.inorder_with_max(root, k)

        return self.values[k - 1]


# @lc code=end

File: rust/src/leetcode/Q230_kth_smallest_element_in_a_bst.rs
/*
 * @lc app=leetcode id=230 lang=rust
 *
 * [230] Kth Smallest Element in a BST
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    fn inorder(root: Option<Rc<RefCell<TreeNode>>>, remained: &mut i32) -> Option<i32> {
        if let Some(node) = root {
            if let Some(result) = Self::inorder(node.borrow().left.clone(), remained) {
                return Some(result);
            }

            *remained -= 1;
            if *remained == 0 {
                return Some(node.borrow().val);
            }

            Self::inorder(node.borrow().right.clone(), remained)
        } else {
            None
        }
    }

    pub fn kth_smallest(root: Option<Rc<RefCell<TreeNode>>>, k: i32) -> i32 {
        let mut remained: i32 = k;
        Self::inorder(root, &mut remained).unwrap()
    }
}
// @lc code=end
